week4/trie/stites/Main.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE InstanceSigs #-}
+--------------------------------------------------------------------------------
+-- 18. Trie /{trie}/
+--
+-- Construct a trie from all words in a dictionary and implement search for
+-- words by prefix. Here's an example of a trie for
+-- /{cool, cat, coal, bet, bean}/:
+--
+--         b       c
+--        /       / \
+--       e       a   o
+--      / \     /   / \
+--     t   a   t   a   o
+--         |       |   |
+--         n       l   l
+--
+-- You should read the words file, construct a trie, say how many nodes are in
+-- the trie (e.g. in the sample one there are 13 nodes), and then answer user's
+-- queries to find all words starting with given letters:
+--
+--     Trie created. There are 13 nodes.
+--
+--     > be
+--     bean bet
+--
+--     > c
+--     cat coal cool
+--
+--     > co
+--     coal cool
+--
+-- You can use the following type for the trie (but feel free to use something
+-- else):
+--
+--     data Trie a = Empty | Node (Map a (Trie a))
+--
+-- The list of words in available in the /data\// folder in the repository.
+--------------------------------------------------------------------------------
+module Main where
+
+import Control.Monad (forever)
+import Data.Monoid
+import Data.List
+import Data.Maybe
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as M
+
+filepath :: FilePath
+filepath = "data/words"
+
+main :: IO ()
+main = do
+  contents <- readFile filepath
+  let t = fromFileContents contents :: Trie Char
+  putStrLn $ "Trie created. There are " ++ show (size t)  ++ " nodes."
```

Comment:
“Trie created” will be printed before the trie is actually created, thanks to laziness. To only print it afterwards, do something like this:

```
let t = fromFileContents contents :: Trie Char
    len = size t
len `seq` putStrLn ("Trie created. There are " ++ show len ++ " nodes.")
```